src/sentry_logs.c

```diff
@@ -0,0 +1,389 @@
+    case SENTRY_LEVEL_INFO:
+        return "info";
+    case SENTRY_LEVEL_WARNING:
+        return "warn";
+    case SENTRY_LEVEL_ERROR:
+        return "error";
+    case SENTRY_LEVEL_FATAL:
+        return "fatal";
+    default:
+        return "unknown";
+    }
+}
+
+static void
+populate_message_parameters(
+    sentry_value_t attributes, const char *message, va_list args)
+{
+    if (!message || sentry_value_is_null(attributes)) {
+        return;
+    }
+
+    const char *fmt_ptr = message;
+    int param_index = 0;
+    va_list args_copy;
+    va_copy(args_copy, args);
+
+    while (*fmt_ptr) {
+        // Find the next format specifier
+        if (*fmt_ptr == '%') {
+            fmt_ptr++; // Skip the '%'
+
+            // Skip flags, width, and precision
+            while (*fmt_ptr
+                && (*fmt_ptr == '-' || *fmt_ptr == '+' || *fmt_ptr == ' '
+                    || *fmt_ptr == '#' || *fmt_ptr == '0')) {
+                fmt_ptr++;
+            }
+            while (*fmt_ptr && (*fmt_ptr >= '0' && *fmt_ptr <= '9')) {
+                fmt_ptr++;
+            }
+            if (*fmt_ptr == '.') {
+                fmt_ptr++;
+                while (*fmt_ptr && (*fmt_ptr >= '0' && *fmt_ptr <= '9')) {
+                    fmt_ptr++;
+                }
+            }
+
+            // Skip length modifiers
+            while (*fmt_ptr
+                && (*fmt_ptr == 'h' || *fmt_ptr == 'l' || *fmt_ptr == 'L'
+                    || *fmt_ptr == 'z' || *fmt_ptr == 'j' || *fmt_ptr == 't')) {
+                fmt_ptr++;
+            }
+
+            if (*fmt_ptr == '%') {
+                // Escaped '%', not a format specifier
+                fmt_ptr++;
+                continue;
+            }
+
+            // Get the conversion specifier
+            char conversion = *fmt_ptr;
+            if (conversion) {
+                char key[64];
+                snprintf(key, sizeof(key), "sentry.message.parameter.%d",
+                    param_index);
+
+                sentry_value_t param_obj = sentry_value_new_object();
+
+                switch (conversion) {
+                case 'd':
+                case 'i': {
+                    int val = va_arg(args_copy, int);
+                    sentry_value_set_by_key(
+                        param_obj, "value", sentry_value_new_int32(val));
+                    sentry_value_set_by_key(
+                        param_obj, "type", sentry_value_new_string("integer"));
+                    break;
+                }
+                case 'u':
+                case 'x':
+                case 'X':
+                case 'o': {
+                    unsigned int val = va_arg(args_copy, unsigned int);
+                    sentry_value_set_by_key(param_obj, "value",
+                        sentry_value_new_int32((int32_t)val));
+                    sentry_value_set_by_key(
+                        param_obj, "type", sentry_value_new_string("integer"));
+                    break;
```

Comment:
We must be cautious here: if you skip the length specifiers, a 64-bit value could be passed in. Even if the value is a 32-bit unsigned, there is no way the backend could know that this is an `int32` containing a `uint32` without an explicit protocol.

Essentially, if we accept unsigned integers of 32-bit, we'd have to store them in a double.

If we accept 64-bit unsigned integers, it must be clear that JSON will overflow numbers at 2^53-1

Reply:
By the way, introducing `sentry_value_t` (a list of kv-objects) as a parameter to format strings would also introduce some normalization in terms of the data that users can pass in here.